_posts/2020-11-03-kafka-commit-strategies.adoc

```diff
@@ -0,0 +1,199 @@
+It commits the offset, indicating that all the previous records from that partition have been processed. 
+So, if a consumer stops and comes back later, it restarts from the last committed position (if assigned to that partition again). 
+Note that this behavior is configurable. 
+
+What's important to notice is the periodic aspect of the commit. 
+Offset commit is expensive, and to enhance performances, we should not commit the offset after each processed record. 
+In this regard, Kafka behaves differently from traditional messaging solutions, such as JMS, which acknowledges each message individually. 
+Another important characteristic is the positional aspect of the commit.
+You commit the position indicating that all the records located before that position are processed. 
+But is it really the case? 
+
+### The Kafka default behavior
+
+The Apache Kafka consumer uses an auto-commit approach by default. 
+Applications using such a consumer are structured around a polling loop:
+
+[source, java]
+----
+while(true) {
+    ConsumerRecords records = consumer.poll(Duration.ofMillis(10000));
+    processRetrievedRecords(records);
+}
+----
+
+Such a program polls a batch of records, processes them, and then polls the next set. 
+While calling the `poll` method, the consumer periodically commits the last offset of the previous batches transparently. 
+
+Quite nice, right? 
+If the application fails to process a message, it throws an exception, which either interrupts the while loop or is handled gracefully (within the `processRetrievedRecords` method). 
+In the first case, it means that it won't commit anymore (as it happens in the `poll` method, not called anymore). 
+If the application restarts, it resumes from the last committed offset (if the consumer is configured with `auto.offset.reset=earliest`).
+It may re-process a set of messages (it's the application's responsibility to handle duplicates), but at least nothing is lost. 
+
+So, is there anything wrong with this?
+Looks wonderful... until you add a pinch of asynchrony.
+
+## What if the message's processing is asynchronous
+
+If the message processing is asynchronous (offloaded to another thread, use non-blocking I/O...), failures may not interrupt the while loop from above. 
+Failure happens asynchronously, outside the polling thread.
+If the `poll` method gets called again despite a failed processing, and _auto-commit_ is still enabled, we may commit offsets while something wrong happened. 
+If some processing of previously retrieved records is not completed yet, while this auto commit happens, it may consider the record as processed correctly, but the outcome is unknown at that point. 
+
+So to handle these case, we can disable the auto-commit and switch to manual commit. 
+In this case, it's the application's responsibility to commit the offsets regularly. 
+So, the application needs to track the polled records, their processing, failures, and periodically commits.
```

Comment:
Should "commits" be plural or singular?

Reply:
It's a verb in this case. I added `commits the offsets`